x-pack/plugins/saved_objects_tagging/server/usage/fetch_tag_usage_data.ts

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { ElasticsearchClient } from 'src/core/server';
+import { TaggingUsageData, ByTypeTaggingUsageData } from './types';
+
+/**
+ * Manual type reflection of the `tagDataAggregations` resulting payload
+ */
+interface AggregatedTagUsageResponseBody {
+  aggregations: {
+    by_type: {
+      buckets: Array<{
+        key: string;
+        doc_count: number;
+        nested_ref: {
+          tag_references: {
+            doc_count: number;
+            tag_id: {
+              buckets: Array<{
+                key: string;
+                doc_count: number;
+              }>;
+            };
+          };
+        };
+      }>;
+    };
+  };
+}
+
+export const fetchTagUsageData = async ({
+  esClient,
+  kibanaIndex,
+}: {
+  esClient: ElasticsearchClient;
+  kibanaIndex: string;
+}): Promise<TaggingUsageData> => {
+  const { body } = await esClient.search<AggregatedTagUsageResponseBody>({
+    index: [kibanaIndex],
+    ignore_unavailable: true,
+    filter_path: 'aggregations',
+    body: {
+      size: 0,
+      query: {
+        bool: {
+          must: [hasTagReferenceClause],
+        },
+      },
+      aggs: tagDataAggregations,
+    },
+  });
```

Comment:
So, ideally we would be using the SO client here. However even when https://github.com/elastic/kibana/pull/64002 lands, I'm not sure we would be able to do so, due to the query I'm using. `hasTagReferenceClause` is filtering to only return the objects that have a reference to a (**any**) tag, by only having a `references.type` constraint. This is something that the SO `find` API does not provide.

As a side-note, this is also what is done for the SO object count in the `kibana` usage collector: https://github.com/elastic/kibana/blob/9ca22382fb9f4aca147e07ac9a42bdb1e9d737e4/src/plugins/kibana_usage_collection/server/collectors/kibana/get_saved_object_counts.ts#L70-L72

Reply:
According to https://github.com/elastic/kibana/issues/69172 it's less efficient to filter an aggregate instead of 'filtering in the query' like you do here. But I wonder if we couldn't use a KQL filter to construct `hasTagReferenceClause`?

Reply:
>  we couldn't use a KQL filter to construct hasTagReferenceClause

AFAIK even with https://github.com/elastic/kibana/pull/64002 we won't be able, as we can't use KQL against top level fields (and current state of https://github.com/elastic/kibana/pull/64002 also forbid aggregations against top level fields)

> According to #69172 it's less efficient to filter an aggregate instead of 'filtering in the query' like you do here

I do both actually. I use

```
term: {
      'references.type': 'tag',
},
```

In the query. However this is not enough: It will returns object that can have reference to both tags, and other objects. This is why I also need to filter in the aggregation before the final bucket aggregation, to avoid including non-tag ids in the `tag_id ` aggregation buckets.

But I'm no expert 😅 . Do you see another way to do that in a single request without the `filter` aggregation?

Reply:
I think I misunderstood https://github.com/elastic/kibana/issues/69172 Seems like a search with a query or filter are both faster than a filter inside an aggregate so it shouldn't matter if we use a KQL filter or a query.

> AFAIK even with #64002 we won't be able, as we can't use KQL against top level fields, such as references

Because of https://github.com/elastic/kibana/issues/81009 validation fails for `type.references` filters (because `references` is a `nested` field), but you can query on `visualization.updated_at` even though it's a root property. I've added this as another reason to https://github.com/elastic/kibana/issues/80912

> This is why I also need to filter in the aggregation before the final bucket aggregation, to avoid including non-tag ids in the tag_id aggregation buckets.

Don't really know how this works either, but I assume this is fairly efficient cause you've already filtered out all irrelevant documents, it's just filtering out the nested values :+1: